### ISSUE TITLE
[glTF 1.0, glTF 2.0] Clarify behavior of parsers when absolute file path URI refs to binary data are present.

### ISSUE THREAD

**Issue by JC3** [KhronosGroup/glTF#1974]
*Edit: It's possible I wasn't clear here. If that is the case, https://github.com/KhronosGroup/glTF/issues/1974#issuecomment-826134473 is an attempt to be more specific.*

In both 1.0 and 2.0 specifications, the behavior of a parser when encountering path-absolute URI references is left open-ended. For example, consider the following file structure:

    /path/
        old/
            duck.gltf
            duck.bin
        new/
            duck.gltf
            duck.bin

Where */path/new/duck.gltf* refers to */path/old/duck.bin*, e.g.:

```
{
   "buffers": [
       {
           "byteLength": 102040,
           "uri": "/path/old/duck.bin"
       }
   ]
}
```

When */path/new/duck.gltf* is loaded, should it load */path/old/duck.bin*, or */path/new/duck.bin*?

What if */path/old/duck.bin* does not exist, but */path/new/duck.bin* does?

The spec states:

> Implementation Note: Clients can optionally support additional URI components. For example http:// or file:// schemes, authorities/hostnames, absolute paths, and query or fragment parameters. Assets containing these additional URI components may be less portable.

That is, it intentionally leaves absolute paths as an optional feature, up to the implementation. However, this creates a situation where an application may write absolute paths, and loader applications tend to be written to be forgiving (due to lack of specification) but the loading behavior ends up application-specific.

Can we get some more well-defined behavior regarding absolute-path references to filesystem URIs? The reliance on the implementation to choose a behavior isn't really necessary but leads to interoperability issues and unexpected situations resulting from ambiguities.

**Comment by lexaknyazev** [comment]
Only when URI starts with `file:` it could be called a "filesystem URI". In that case, applications should refer to [RFC 8089](https://tools.ietf.org/html/rfc8089).

When the scheme is not set explicitly, it should first be derived from the runtime context (e.g. `file:` for local usage or `http:` for the web) and then an application should follow the appropriate resolution process.

/cc @bghgary

**Comment by JC3** [comment]
Hi. I'm unclear on how that is relevant. Everything you say *is* correct, but was also never called into question (although I've edited the title to be slightly more technically correct). :)

There are no issues regarding the syntax of a URI reference, and there are no questions regarding how to resolve URI paths. There is no requirement for a `file:` scheme (or any scheme) to appear in URI references ([RFC 3986 4.2](https://tools.ietf.org/html/rfc3986#section-4.2)), and URI references starting with a `/` are termed "path-absolute" by the URI RFC. 

There are also no issues regarding the use of the phrase "filesystem URI". As you imply, URI reference resolution is done relative to a base URI, in the case of glTF the base URI is ultimately determined as described in RFC 3986 5.1.2 (base URI = URI of encapsulating entity) -- the glTF spec does not need to be explicit about this, as it is part the URI spec. Since my example refers to filesystem objects, the base URI of the file */path/new/duck.gltf* is necessarily `file:///path/new/duck.gltf`, and therefore all of the paths and URI references I mention end up with the `file:` scheme if you really want to resolve them to URIs, and I do not need to explicitly mention the scheme: Every URI reference I mention resolves to a filesystem URI, basically because I said they did. :)

The glTF specification does *not* require support for anything besides `data:` URIs and a subset of relative URI references ("path-noscheme", but not "path-absolute"), and explicitly states this in the 2.0 spec:

> Clients must support at least these two URI types:
> - Data URIs that embed resources in the JSON. They use syntax defined by RFC 2397.
> - Relative URI paths — or path-noscheme as defined by RFC 3986, Section 4.2 — without scheme, authority, or parameters.

No URI-related RFCs specify what an application should do with a URI after applying the appropriate resolution process, nor do any define application behavior when presented with a situation that is technically unambiguous but contextually (and realistically) ambiguous, such as my example. 

--- 

# Issue Details

All that said, I'll try to be more specific:

A URI reference is define (by [RFC 3986 4.1](https://tools.ietf.org/html/rfc3986#section-4.1)) as:

    URI-reference = URI / relative-ref

A relative reference is defined as:

    relative-ref  = relative-part [ "?" query ] [ "#" fragment ]

    relative-part = "//" authority path-abempty
                    / path-absolute
                    / path-noscheme
                    / path-empty

And the main difference between "path-absolute" and "path-noscheme" is that "path-noscheme" can't start with a `/`. In the context of filesystem objects, "path-absolute" is an absolute file path and "path-noscheme" is a relative file path.

The problem is, for reasons that I'm assuming made sense at the time, but are a mystery to me, **the glTF specification chose to only *require* support for the "path-noscheme" subset of URI relative references, *and* rather than forbidding or requiring "path-absolute" support, it just made it optional**.

Regardless of how much / little sense one thinks this makes, it leads to a number of hard-to-avoid issues in the context of file systems when reality is taken into account:

First of all, you literally have to put extra effort into making an application that supports relative file paths also *not* support absolute file paths. Every file IO subsystem I can think of just handles absolute and relative path resolution transparently. What this means is that, unless an application specifically parses the URI reference at the application layer (which it almost never does) then detects and rejects absolute file paths, it can end up partially supporting an optional section of the glTF spec *even if it did not explicitly intend to*. The end result is it may appear to support part of the spec but in reality said support may not be well-defined in the application (this may or may not have consequences).

More importantly, consider my example: Technically, an application loading */path/new/duck.gltf* should look for */path/old/duck.bin*. Realistically, we can see that that almost certainly isn't intended, and easily leads to a situation such as an application writing an absolute path (which, again, it is allowed to do as per glTF) and then a user moving the file or transferring it to somebody else, thus breaking the path.

A more confusing situation arises if a user *copied* the file to somewhere else (leading to something like my example), they can easily end up in a situation where an application that recognizes absolute paths (i.e. an application that *doesn't go through the extra effort of **rejecting** absolute paths* -- which the glTF standard currently makes you do) appears to load */path/new/duck.gtlf* correctly, until they delete the seemingly unrelated */path/old/* and the application confusingly stops being able to load the file they copied (or, even worse, perhaps they make changes to the copy but observe no changes reflected in an application that loads it).

So then, because these are such common situations, the *reality* is application developers have to be a bit loose about what to do when encountering absolute paths that either don't exist, or appear to be unintended: The priority in most application development is functionality and user satisfaction. If I send you *duck.gltf* + *duck.bin* and it contains an absolute path, in context they clearly go together, and you don't care if the absolute path is incorrect: you care about loading the file. This is what application developers work towards.

It doesn't matter that they should technically reject absolute file paths that don't exist (or do exist but might not make sense); when said absolute file paths end up being encountered in the wild and the *intended* target of the URI reference is clear, most developers are going to make their application work with that, either due to user complaints or to their own desire to maximize functionality and interoperability. A technical specification, then, should work towards eliminating ambiguity arising from realistic scenarios.

---

# Conclusion / TL;DR

glTF requires support for the "path-noscheme" subset of URI relative references, but leaves "path-absolute" optional. When taken in the context of filesystem paths, it creates a number of ambiguities.

Therefore, at the end of the day, you have a situation where lack of explicitness in a standard leads to applications taking liberties in their interpretation of said standard. And, while it may appear at first glance that some of the example situations I just mentioned are high-level application issues, outside of the scope of glTF, the reality is that the are all direct consequences of glTF's underspecified handling of absolute paths, which is precisely the purpose of having standards and specifications to begin with. 

And so I request that:

- The glTF standard be explicit about the details of "path-absolute" handling, or at *least* provide some sort of strong guidelines.
- Future glTF standards should either *disallow* absolute file paths, or *require* them. Requiring relative filesystem path support but leaving absolute filesystem path support as an option opens the door to too many ambiguous situations.

That's all I'm asking. That was a lot of words for a straightforward request, I guess. 😂

**Comment by JC3** [comment]
PS Incidentally, on the subject of correctness, the glTF spec states that it uses "URIs", but really what it uses is "Data-scheme URIs and a subset of URI relative references". If we want to strictly match RFC 3986 terminology, "URIs" is a superset of what glTF uses. This is more an observation than an issue. I think it's clear enough in context, anyways. 🤷‍♂️

**Comment by lexaknyazev** [comment]
> More importantly, consider my example: Technically, an application loading _/path/new/duck.gltf_ should look for _/path/old/duck.bin_.

In that example, nothing suggests that an app should selectively rewrite _/path/old/duck.bin_ to _/path/new/duck.bin_.

Encountering a `path-absolute` may be confusing but neither requiring nor disallowing it would help here.
- Some implementations will never support absolute paths for security reasons, they may even reject path segments like `../`.
- Other implementations may reasonably use them in virtualized contexts (e.g. when the "root" is app-defined).

The current note says that such assets are less portable, i.e. exporters should avoid absolute paths in assets that are intended to be distributed externally. I think we could call out that issue more explicitly.

**Comment by JC3** [comment]
> > More importantly, consider my example: Technically, an application loading _/path/new/duck.gltf_ should look for _/path/old/duck.bin_.
> 
> In that example, nothing suggests that an app should selectively rewrite _/path/old/duck.bin_ to _/path/new/duck.bin_.

Perhaps. But consider that example with this bit of added information: The user in that example copied _/path/old/*_ to _/path/new/*_. With that extra info, there's no technical reason to selectively rewrite the path, but the *intent* is clear *and* nothing violated the glTF specs. (Plus I think I raised this scenario sufficiently in other examples.)

If glTF explicitly forbade absolute file paths, then the application that wrote that file would not have had the option to include the absolute path, and no unexpected situation would occur. And, to a lesser extent, any application that read that file would, in order to be compliant with the spec, have to fail even if the file existed, which would immediately highlight the issue.

If glTF explicitly required absolute file path support and defined the behavior, then the application that read that file would have to fail if the file didn't exist, there would be no need to second-guess, and the issue would be highlighted immediately. To a lesser extent, any standards-conscious developer of an application that writes files would hopefully be aware of the implications of choosing an absolute file path, possibly being encouraged to expose it as a settable option, or whatever, etc.

Either way, the situation essentially disappears because clarity doesn't put developers in a situation that encourages them to guess what the user intended.

Plus, it's a relatively straightforward way to remove a ton of interoperability issues.

I don't feel that the inclusion of an interoperability warning in a specification should preclude said specification from attempting to mitigate interoperability issues where it can.

All that said, I acknowledge that these messy situations are already out in the wild and have been for a while, and that it might be too late. It would be unlikely for developers of existing applications to notice any change in the spec, and it would be unlikely (for reasons given) for them to sacrifice the existing flexibility of their application by e.g. suddenly refusing to load files with absolute file paths. So it be too late for 2.0, but it really ought to be in a future version, if there is one.

**Comment by JC3** [comment]
Oh; adding:

> The current note says that such assets are less portable, i.e. exporters should avoid absolute paths in assets that are intended to be distributed externally. 

You know as well as I do that "warning: less portable" doesn't really get anybody anywhere outside of technical arguments. :) As soon as a "non-portable" asset is encountered, almost any developer is going to try to find a way to handle it, *especially* if their software has a large user base or frequently encounters said assets. That's why specs should account for those situations when possible.

> I think we could call out that issue more explicitly.

I agree. It could help a bit; especially for new implementors.

**Comment by emackey** [comment]
One of the above comments mentions security, but I'll bring it up again here because I feel it's being overlooked.  Many desktop applications today have an even tighter focus on security than they do on enabling all the loose-ends features under discussion here.

As a quick test, I copied the Duck.gltf model to a new folder (its own screenshot folder), and renamed the paths to say `../glTF/Duck0.bin` and `../glTF/DuckCM.png`.  This is intentionally broken in many desktop apps, not because of anything written in the spec, but because "Duck.gltf" came from the filesystem and the security model can't trust it to access the parent folder or siblings.  Office won't load it, 3D Viewer won't touch it, and my VSCode extension won't display it (because its webview must supply a list of authorized folders to access, and siblings aren't included).

On the web it's a different story, because the http/https protocol provides a security model.  One could use `/path/foo.bin` on a glTF file that came from the https protocol, and understand that the leading slash just gets you to the root of the "same origin", so the security is fine.  One could even pull texture images from absolute URIs to a remote CDN via CORS I suppose.  We don't have any reason to forbid absolute `https` URIs, particularly for web-hosted apps, but we also can't demand that desktop apps honor network URIs.

Long story short, I don't think this is much of an issue, because applications are more strongly constrained by the security concerns than by what's written in the glTF spec.  We can't make assumptions about every other application's security model, how much network access it needs or prevents, and how much filesystem access it needs or prevents.

This is an area where it might be better to leave the support optional, as the spec is now, and let apps adapt it to their own security models.  I don't think there's high risk of exporters polluting the ecosystem with absolute paths.  None of the popular exporters write absolute paths now, and I think it's well-understood that desktop apps aren't going to let glTFs run wild across the filesystem.

**Comment by lexaknyazev** [comment]
Just like HTML or SVG documents, a glTF asset with absolute paths may behave differently depending on the execution context or the application security model. Neither of these are in scope of the glTF spec.
